walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,110 @@
+
+    Collection<TimeRange> acceptableMandatoryTimeRanges = getConflictFreeTimeRanges(
+        mandatoryAttendeeTimeRanges, request);
+    Collection<TimeRange> acceptableOptionalTimeRanges = getConflictFreeTimeRanges(
+        allAttendeeTimeRanges, request);
+
+    if (request.getAttendees().isEmpty() || !acceptableOptionalTimeRanges.isEmpty()) {
+      return acceptableOptionalTimeRanges;
+    } else {
+      return acceptableMandatoryTimeRanges;
+    }
+
+  }
+
+  private Collection<TimeRange> getConflictFreeTimeRanges(
+      Collection<TimeRange> timeRanges, MeetingRequest request) {
+    // add dummy events at start of day and end of day with no duration 
+    timeRanges.add(START_OF_DAY_TIMERANGE);
+    timeRanges.add(END_OF_DAY_TIMERANGE);
+    
+    List<TimeRange> timeRangesByStart = timeRanges.stream()
+        .sorted(TimeRange.ORDER_BY_START)
+        .collect(Collectors.toList());
+
+    List<TimeRange> acceptableTimeRanges = new ArrayList<TimeRange>();
+    
+    // edge case: if meeting request duration is too long, return empty list
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return acceptableTimeRanges;
+    }
+
+    // edge case: if there are no exisiting meetings, return entire day
+    if (timeRanges.isEmpty()) {
+      acceptableTimeRanges.add(TimeRange.WHOLE_DAY);
+      return acceptableTimeRanges;
+    }
+
+    TimeRange currentMeetingTimeRange;
+    TimeRange nextMeetingTimeRange;
+    for (int i = 0; i < timeRangesByStart.size() - 1; i++) {
+      currentMeetingTimeRange = timeRangesByStart.get(i);  
+      nextMeetingTimeRange = timeRangesByStart.get(i+1);
+      // if overlap, merge with next meeting
+      if (currentMeetingTimeRange.overlaps(nextMeetingTimeRange)) { 
+          int newEnd = Math.max(currentMeetingTimeRange.end(), nextMeetingTimeRange.end());
+          timeRangesByStart.set(i+1, TimeRange.fromStartEnd(
+              currentMeetingTimeRange.start(), newEnd, /* exclusive of end */ false));
+      }
+      else { 
+        if (canGapFitRequest(currentMeetingTimeRange.end(), nextMeetingTimeRange.start(), request)) {
+          TimeRange accetableRange = TimeRange.fromStartEnd(
+              currentMeetingTimeRange.end(), nextMeetingTimeRange.start(), 
+              /* not inclusive of end */ false);
+          acceptableTimeRanges.add(accetableRange);
+        }
+      }
+    }
+
+  return acceptableTimeRanges;
+  }
+
+  private Collection<TimeRange> extractAttendingTimeRangesFromEvents(
+      Collection<Event> events, MeetingRequest request, boolean includeOptionalAttendees) {   
+    Collection<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    Collection<String> requestedAttendees = request.getAttendees();
+
+    if (includeOptionalAttendees) {
+      requestedAttendees = Stream.of(requestedAttendees, request.getOptionalAttendees())
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
```

Comment:
Funky Indents here

Reply:
I tried to follow +2 spaces for block indentation and +4 spaces for line indentation according to the styleguide https://google.github.io/styleguide/javaguide.html